### PR TITLE
Added `LambdaExpressionVisitor` and new `ExpressionOptions` from NCalc2

### LINKED ISSUE #185: Create `LambdaExpressionVisitor`
Suggested by @Bykiev. I see no problem. Should this be customized in a interface like `ParameterExtractionVisitor`?

### LINKED ISSUE #189: Add more options from NCalc2
```
  // Summary:
        //     Use ordinal culture on string compare
        MatchStringsOrdinal = 1 << 6,
        //
        // Summary:
        //     Use checked math
        OverflowProtection = 1 << 7,

        /// <summary>
        ///     Allow calculation with boolean values.
        /// </summary>
        BooleanCalculation = 1 << 8,

        /// <summary>
        ///     When using Abs(), return a double instead of a decimal.
        /// </summary>
        UseDoubleForAbsFunction = 1 << 9,

        /// <summary>
        /// Defines a "null" parameter and allows comparison of values to null.
        /// </summary>
        AllowNullParameter = 1 << 10
```

### LINKED ISSUE #190: Regression bug - OverflowException with double values
I've found that in `LogicalExpressionParser` the result variable type is `decimal`, but I think it should be `double` and only when `ExpressionOptions.DecimalAsDefault` (`UseDecimalsAsDefault`) specified it should be converted to decimal. 

https://github.com/ncalc/ncalc/blob/e6c8c5f47932810a9f2f00397a95d31d058b4722/src/NCalc/Parser/LogicalExpressionParser.cs#L88-L95

Because of this such expression cause overflow:

```
var expr = new Expression($"Floor({double.MaxValue.ToString(CultureInfo.InvariantCulture)})");
var res = expr.Evaluate();
```

Seems to be a regression bug since 4.0

### PR DESCRIPTION
**Pull request by gumbarros**
This closes #185 
This closes #190 
This closes #189 

Credits to @sklose for the original implementation.

New ExpressionOptions:
`AllowBooleanCalculation`
`OrdinalStringComparer`

Co-authored-by: Sebastian Klose
[mail@sklose.com](mailto:mail@sklose.com)

Co-authored-by: Andrew Bykiev
[Bykiev](https://github.com/Bykiev)

### CODE REVIEW

**Comment by gumbarros** [comment]
Help is wanted at ExpressionAndLambdaFuncBehaviorMatch test

**Comment by Bykiev** [comment]
`ExpressionOptions.BooleanCalculation` seems to be missing

**Comment by Bykiev** [comment]
I think it's a parser bug with integer overflow, just check this simple expression to get the idea:

```
var expr = new Expression($"Abs({decimal.MinValue})", CultureInfo.InvariantCulture);
var res = expr.Evaluate();
```

**Comment by gumbarros** [comment]
> `ExpressionOptions.BooleanCalculation` seems to be missing

I Will track at #189 in another PR with CheckedMath, sees to not causes any errors with lambda

**Comment by gumbarros** [comment]
@Bykiev ready to merge

**Comment by gumbarros** [comment]
@Bykiev I also added your fixes from #191

File: src/NCalc/Parser/LogicalExpressionParser.cs
Comment (Bykiev) [inline src/NCalc/Parser/LogicalExpressionParser.cs:209]: Should we pass `NumberOptions.AllowSign` here?

**Comment by Bykiev** [comment]
Though it doesn't related to this PR, but this expression is still failing both in .NET Framework and .NET:

```
var expr = new Expression($"Floor(12e+100)");
var res = expr.Evaluate();

Assert.Equal(Math.Floor(12e+100), res);
```

Seems due to how floating point values are stored in memory, I think it can be resolved only by using `BigDecimal` implementation. 

I think it should be another PR.

**Comment by Bykiev** [comment]
We should add docs and update the changelog before merging

**Comment by Bykiev** [comment]
I'm confused with `ExpressionOptions.UseDoubleForAbsFunction` and `ExpressionOptions.DecimalAsDefault` options. Do we really need `ExpressionOptions.UseDoubleForAbsFunction` or we can use `ExpressionOptions.DecimalAsDefault` value (so, by default `Abs` will return double and if `ExpressionOptions.DecimalAsDefault` specified the decimal will be returned)?

**Comment by gumbarros** [comment]
> I'm confused with `ExpressionOptions.UseDoubleForAbsFunction` and `ExpressionOptions.DecimalAsDefault` options. Do we really need `ExpressionOptions.UseDoubleForAbsFunction` or we can use `ExpressionOptions.DecimalAsDefault` value (so, by default `Abs` will return double and if `ExpressionOptions.DecimalAsDefault` specified the decimal will be returned)?
> 
> Or maybe we need to rename it to `UseDecimalForAbsFunction` as other functions returns double?

You're right, I was also confused when I added the option. Maybe they created this option because `ExpressionOptions.DecimalAsDefault` does not exists at NCalc2, makes sense use this.

**Comment by gumbarros** [comment]
image

Just removing the option cause this.
I vote in `UseDecimalForAbsFunction`, do you agree?

**Comment by gumbarros** [comment]
My suggestion, if DecimalAsDefault, all functions should expect a decimal as parameter

**Comment by Bykiev** [comment]
> > I'm confused with `ExpressionOptions.UseDoubleForAbsFunction` and `ExpressionOptions.DecimalAsDefault` options. Do we really need `ExpressionOptions.UseDoubleForAbsFunction` or we can use `ExpressionOptions.DecimalAsDefault` value (so, by default `Abs` will return double and if `ExpressionOptions.DecimalAsDefault` specified the decimal will be returned)?
> > Or maybe we need to rename it to `UseDecimalForAbsFunction` as other functions returns double?
> 
> You're right, I was also confused when I added the option. Maybe they created this option because `ExpressionOptions.DecimalAsDefault` does not exists at NCalc2, makes sense use this.

I think so, that's why we should adopt it to this branch

**Comment by Bykiev** [comment]
> My suggestion, if DecimalAsDefault, all functions should expect a decimal as parameter

Agreed

**Comment by gumbarros** [comment]
@Bykiev if you want to make any changes in this branch, you are a Member of this repo

**Comment by Bykiev** [comment]
This release is full of breaking changes, but all of them are needed to make NCalc better

**Comment by gumbarros** [comment]
> This release is full of breaking changes, but all of them are needed to make NCalc better

Yeah and without them is impossible to finish Lambda support

**Comment by Bykiev** [comment]
I've tried to find the purpose of `EvaluateOptions.UseDoubleForAbsFunction` in NCalc2, but didn't find anything except this PR https://github.com/sklose/NCalc2/pull/31. I've tried the latest version and there is no any errors while evaluating expression without this option. Seems a lot has been changed since this PR and now this option seems very strange

**Comment by gumbarros** [comment]
> I've tried to find the purpose of `EvaluateOptions.UseDoubleForAbsFunction` in NCalc2, but didn't find anything except this PR [sklose/NCalc2#31](https://github.com/sklose/NCalc2/pull/31). I've tried the latest version and there is no any errors while evaluating expression without this option. Seems a lot has been changed since this PR and now this option seems very strange

I understand the purpose, it's because some Math functions can have both double and decimals as parameter.

**Comment by gumbarros** [comment]
@Bykiev it's holiday here and I'm going to be unavailable until Monday night, but don't hesitate to tag me for any questions. Thanks for the help these days!

**Comment by Bykiev** [comment]
> @Bykiev it's holiday here and I'm going to be unavailable until Monday night, but don't hesitate to tag me for any questions. Thanks for the help these days!

Have a nice holiday!

**Comment by Bykiev** [comment]
@gumbarros, @sklose, @david-brink-talogy, can you please review it?

File: docs/articles/lambda_compilation.md
Comment (gumbarros) [inline docs/articles/lambda_compilation.md:3]: Question about perf, it's still faster than Evaluate with the Compilation or it's recommended at the docs to say the consumer to cache the lambda?

File: src/NCalc/Visitors/LambdaExpressionVisitor.cs
Comment (gumbarros) [inline src/NCalc/Visitors/LambdaExpressionVisitor.cs:394]: If this variable never changes, we have move it to a `HashSet<string>` at `TypeHelper` right?

File: docs/articles/lambda_compilation.md
Comment (gumbarros) [reply docs/articles/lambda_compilation.md:3]: I can also later at MemoryCache package to add a option to cache the lambdas if the Compilation+MethodCall is slower than Evaluate, but this should be a separated PR I think

Comment (Bykiev) [reply docs/articles/lambda_compilation.md:3]: Good idea, but we've already have an article about caching. Not sure about performance with a cached lambda

File: src/NCalc/Visitors/LambdaExpressionVisitor.cs
Comment (Bykiev) [reply src/NCalc/Visitors/LambdaExpressionVisitor.cs:394]: As we need an order the HashSet is not the best way as the index in HashSet is meaningless

Comment (gumbarros) [reply src/NCalc/Visitors/LambdaExpressionVisitor.cs:394]: @Bykiev agree, based on NCalc2 I changed the arrays of TypeHelper to HashSet, can you revert please?

Comment (Bykiev) [reply src/NCalc/Visitors/LambdaExpressionVisitor.cs]: Added the changes in https://github.com/ncalc/ncalc/pull/188/commits/d29d74bd7b9cb14621339f809b42cca81f5f41f8

Comment (Bykiev) [reply src/NCalc/Visitors/LambdaExpressionVisitor.cs]: I don't like the logic of `GetMostPreciseNumberType`, should we make it the same as `GetMostPreciseType`?

Comment (gumbarros) [reply src/NCalc/Visitors/LambdaExpressionVisitor.cs]: I am on the phone and can't see much, but I agree in the lambda evaluation to re-use the most possible of our original visitor

Comment (Bykiev) [reply src/NCalc/Visitors/LambdaExpressionVisitor.cs]: The logic is quite different, I can't say for now if such change is ok. Let's leave it as is.

**Review by Bykiev — APPROVED** [review]

File: CHANGELOG.md
Comment (gumbarros) [inline CHANGELOG.md:5]: It's not obsolete and non breaking, I just added CaseInsensitiveStringComparer to match the naming with OrdinalStringComparer. CaseInsensitiveComparer still exists but with obsolete attribute

Comment (Bykiev) [reply CHANGELOG.md:5]: Sorry, don't understand which change about you are not agreed

Comment (Bykiev) [reply CHANGELOG.md:5]: The breaking change section is below

Comment (gumbarros) [reply CHANGELOG.md:5]: OrdinalStringComparer is not related to CaseInsensitiveComparer, is a totally new feature

Comment (Bykiev) [reply CHANGELOG.md:5]: Thanks, should be CaseInsensitiveStringComparer, but it's not in a breaking change section

**Comment by gumbarros** [comment]
@Bykiev I think this is ready to merge, but I vote in a v5 after a few PRs. I want to improve error handling again (request at Discord server, will open issue) .

**Comment by Bykiev** [comment]
> @Bykiev I think this is ready to merge, but I vote in a v5 after a few PRs. I want to improve error handling again (request at Discord server, will open issue) .

Do you really think we should update the major version?

**Review by sklose — APPROVED** [review]
lgtm, great job merging the two projects @Bykiev and @gumbarros  🚀

**Comment by gumbarros** [comment]
> > @Bykiev I think this is ready to merge, but I vote in a v5 after a few PRs. I want to improve error handling again (request at Discord server, will open issue) .
> 
> Do you really think we should update the major version?

I'm in doubt because of semver lots of breaking changes, but thinking again, they are runtime changes, and semver is not a universal rule, changing my vote to 4.2 again haha

**Comment by Bykiev** [comment]
Then, let's merge it!

**Pull request merged**
